service/history/shard/controller_impl.go

```diff
@@ -446,12 +464,63 @@ func (c *ControllerImpl) acquireShards(ctx context.Context) {
 	_ = sem.Acquire(ctx, concurrency)
 
 	c.RLock()
+
 	numOfOwnedShards := len(c.historyShards)
+	// Readiness check: We should set initialShardsAcquired when:
+	// 1. It's not already set.
+	// 2. We should own at least one shard (i.e. not before we join membership).
+	// 3. We have ownership of all the shards we're supposed to own.
+	if !c.initialShardsAcquired.Ready() && numOfOwnedShards > 0 {
+		go c.checkShardReadiness(readinessCtx, readinessCancel, expmaps.Values(c.historyShards))
+	} else {
+		readinessCancel()
+	}
+
 	c.RUnlock()
 	metrics.NumShardsGauge.With(c.taggedMetricsHandler).Record(float64(numOfOwnedShards))
 	c.publishShardCountUpdate(numOfOwnedShards)
 }
 
+func (c *ControllerImpl) checkShardReadiness(
+	ctx context.Context,
+	cancel context.CancelFunc,
+	shards []historyi.ControllableContext,
+) {
+	defer cancel()
+
+	concurrency := int64(max(c.config.AcquireShardConcurrency(), 1))
+	sem := semaphore.NewWeighted(concurrency)
+	var ready atomic.Int32
+	for _, shard := range shards {
+		if sem.Acquire(ctx, 1) != nil {
+			return
```

Comment:
It's possible to return from this function while a subset of go routines below are running. Though, perhaps since each of the potentially blocking functions in the goroutine share this same context it's probably okay to assume they'll return due to cancelation?

Reply:
Yes and no... I originally tried to use contexts to get everything to clean up neatly, but it turns out AssertOwnership uses a detached context. So we can't actually interrupt it, but that's fine, the goroutine isn't doing anything interesting after it completes